File: backend/firebase.py
import pyrebase
from dotenv import load_dotenv

import os

load_dotenv()

config = {
    "apiKey": os.getenv('FIREBASE_API_KEY'),
    "authDomain": os.getenv('AUTH_DOMAIN'),
    "databaseURL": os.getenv('DATABASE_URL'),
    "projectId": os.getenv('PROJECT_ID'),
    "storageBucket": os.getenv('STORAGE_BUCKET'),
    "serviceAccount": os.getenv('SERVICE_ACCOUNT'),
    "messagingSenderId": os.getenv('MESSAGING_SENDER_ID'),
    "appId": os.getenv('APP_ID')
}


class Firebase:
    def __init__(self, config):
        self.firebase = pyrebase.initialize_app(config)
        self.db = self.firebase.database()

    def add(self, key, item):
        self.db.child(key).push(item)

    def set(self, key, iid, item):
        self.db.child(key).child(iid).set(item)

    def get(self, key):
        return self.db.child(key).get().val().values()

    def update(self, key, iid, item):
        self.db.child(key).child(iid).update(item)

    def remove(self, key, iid):
        self.db.child(key).child(iid).remove()


firebase = Firebase(config)

File: frontend/src/components/Firebase/firebase.js
import app from 'firebase/app';
import 'firebase/auth';
import 'firebase/database';

const config = {
    apiKey: process.env.REACT_APP_API_KEY,
    authDomain: process.env.REACT_APP_AUTH_DOMAIN,
    databaseURL: process.env.REACT_APP_DATABASE_URL,
    projectId: process.env.REACT_APP_PROJECT_ID,
    storageBucket: process.env.REACT_APP_STORAGE_BUCKET,
    messagingSenderId: process.env.REACT_APP_MESSAGING_SENDER_ID,
    appId: process.env.REACT_APP_APP_ID
};

class Firebase {
    constructor() {
        app.initializeApp(config);

        this.auth = app.auth();
        this.db = app.database();
    }

    /* AUTH API */

    doCreateUserWithEmailAndPassword = (email, password) =>
        this.auth.createUserWithEmailAndPassword(email, password);

    doSignInWithEmailAndPassword = (email, password) =>
        this.auth.signInWithEmailAndPassword(email, password);

    doSignOut = () =>
        this.auth.signOut();

    doPasswordReset = email =>
        this.auth.sendPasswordResetEmail(email);

    doPasswordUpdate = password =>
        this.auth.currentUser.updatePassword(password);

    /* MERGE AUTH AND DB USER INFO */

    onAuthUserListener = (next, fallback) =>
        this.auth.onAuthStateChanged(authUser => {
            if (authUser) {
                this.user(authUser.uid)
                    .once('value')
                    .then(snapshot => {
                        const dbUser = snapshot.val();

                        if (!dbUser.roles) {
                            dbUser.roles = {};
                        }

                        // merge
                        authUser = {
                            uid: authUser.uid,
                            email: authUser.email,
                            ...dbUser
                        };

                        next(authUser);
                    });
            } else {
                fallback();
            }
        });

    /* USER API */

    user = uid => this.db.ref(`users/${uid}`);

    users = () => this.db.ref('users');

    /* VEHICLE API */

    vehicle = vid => this.db.ref(`vehicles/${vid}`);

    vehicles = () => this.db.ref('vehicles');
}

export default Firebase;